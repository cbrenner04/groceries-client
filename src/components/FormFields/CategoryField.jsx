import React from 'react';
import PropTypes from 'prop-types';

const CategoryField = props => (
  <div className="form-group">
    <label htmlFor="category">Category</label>
    <input
      name="category"
      type="text"
      className="form-control"
      id="category"
      value={props.category}
      onChange={props.handleInput}
      list="categories"
    />
    <datalist id="categories">
      {props.categories.map(category => <option key={category} value={category} />)}
    </datalist>
  </div>
);

CategoryField.propTypes = {
  handleInput: PropTypes.func.isRequired,
  category: PropTypes.string,
  categories: PropTypes.arrayOf(PropTypes.string),
};

CategoryField.defaultProps = {
  category: '',
  categories: [],
};

export default CategoryField;
