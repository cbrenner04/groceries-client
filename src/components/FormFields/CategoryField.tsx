import React, { type ChangeEventHandler, type ReactNode } from 'react';
import { Form } from 'react-bootstrap';

export interface ICategoryFieldProps {
  handleInput: ChangeEventHandler;
  category?: string;
  categories?: string[];
  name?: string;
  child?: ReactNode;
  disabled?: boolean;
}

const CategoryField: React.FC<ICategoryFieldProps> = ({
  category = '',
  categories = [],
  handleInput,
  name = 'category',
  child = '',
  disabled = false,
}): React.JSX.Element => (
  <Form.Group controlId={name} className="mb-3">
    <Form.Label>Category</Form.Label>
    <Form.Control
      type="text"
      value={category}
      onChange={handleInput}
      list="categories"
      name={name}
      disabled={disabled}
    />
    <datalist id="categories">
      {categories.map((category) => (
        <option key={category} value={category} />
      ))}
    </datalist>
    {child}
  </Form.Group>
);

export default CategoryField;
