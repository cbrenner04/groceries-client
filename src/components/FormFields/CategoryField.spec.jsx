import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import CategoryField from './CategoryField';

const defaultProps = {
  handleInput: jest.fn(),
  category: 'testCategory',
  categories: ['testCategory1'],
};

describe('CategoryField', () => {
  it('renders input with datalist', () => {
    const { getByLabelText } = render(<CategoryField {...defaultProps} />);
    const formInput = getByLabelText('Category');
    const formGroup = formInput.parentElement;

    expect(formGroup).toMatchSnapshot();
    expect(formInput).toHaveValue(defaultProps.category);
    expect(formGroup.children[2].firstChild).toHaveAttribute('value', defaultProps.categories[0]);
  });

  describe('when value changes', () => {
    it('calls handleInput', () => {
      const { getByLabelText } = render(<CategoryField {...defaultProps} />);
      const formInput = getByLabelText('Category');
      fireEvent.change(formInput, { target: { value: 'a' } });

      expect(defaultProps.handleInput).toHaveBeenCalled();
    });
  });
});
