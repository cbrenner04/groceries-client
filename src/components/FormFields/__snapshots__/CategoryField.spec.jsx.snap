// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryField renders input with datalist 1`] = `
<div
  class="mb-3"
>
  <label
    class="form-label"
    for="category"
  >
    Category
  </label>
  <input
    class="form-control"
    id="category"
    list="categories"
    name="category"
    type="text"
    value="testCategory"
  />
  <datalist
    id="categories"
  >
    <option
      value="testCategory1"
    />
  </datalist>
</div>
`;
